pkg/planner/core/operator/logicalop/logical_join.go

```diff
@@ -1263,26 +1268,43 @@ func (p *LogicalJoin) PreferAny(joinFlags ...uint) bool {
 	return false
 }
 
-// canPropagateConstantWithInnerJoinOrSemiJoin is to It is used to determine whether the newly created expression
-// during constant propagation can be pushed down to the child nodes. If the new expression cannot be pushed down,
-// we will remove it.
 // This function is only used with inner join and semi join.
-func (p *LogicalJoin) canPropagateConstantWithInnerJoinOrSemiJoin(expr expression.Expression) bool {
-	_, _, _, otherCond := p.extractOnCondition([]expression.Expression{expr}, true, true)
-	// If otherCond is empty, we may consider this expression to be useful.
-	// This expression is created by the constant propagation
-	return len(otherCond) == 0
+func (p *LogicalJoin) isVaildConstantPropagationExpressionWithInnerJoinOrSemiJoin(expr expression.Expression) bool {
+	return p.isVaildConstantPropagationExpression(expr, true, true)
 }
 
-// canPropagateConstantForJoinPropConst is to It is used to determine whether the newly created expression
-// during constant propagation can be pushed down to the child nodes. If the new expression cannot be pushed down,
-// we will remove it.
 // This function is only used in LogicalJoin.joinPropConst.
-func (p *LogicalJoin) canPropagateConstantForJoinPropConst(expr expression.Expression) bool {
-	_, _, _, otherCond := p.extractOnCondition([]expression.Expression{expr}, false, false)
-	// If otherCond is empty, we may consider this expression to be useful.
-	// This expression is created by the constant propagation
-	return len(otherCond) == 0
+func (p *LogicalJoin) isVaildConstantPropagationExpressionForJoinPropConst(expr expression.Expression) bool {
+	return p.isVaildConstantPropagationExpression(expr, false, false)
+}
+
+// isVaildConstantPropagationExpression is to judge whether the expression is created by PropagationContant is vaild.
+//
+// Some expressions are not suitable for constant propagation. After constant propagation,
+// these expressions will only become a projection, increasing the computational load without
+// being able to filter data directly from the data source.
+func (p *LogicalJoin) isVaildConstantPropagationExpression(cond expression.Expression, deriveLeft bool, deriveRight bool) bool {
+	_, leftCond, rightCond, otherCond := p.extractOnCondition([]expression.Expression{cond}, deriveLeft, deriveRight)
+	if len(otherCond) > 0 {
+		// a new expression which is created by constant propagation, is a other condtion, we don't put it
+		// into our final result.
+		return false
+	}
+	// When the expression is a left/right condition, we want it to filter more of the underlying data.
+	if len(leftCond) > 0 || len(rightCond) > 0 {
+		colset := expression.ExtractColumnsUniqueIDFromExpressions(cond)
```

Comment:
Almost LGTM, but could we merge `ExtractColumnsUniqueIDFromExpressions` and `isAllUniqueIDInTheSameTable` into one single function like `ExpressionColumnsFromTheSameTable`?

Reply:
I have refactored it.